### PR TITLE
add support for binding symfony command

### LINKED ISSUE #209: SymfonyCommandTask: wrap a console command in castor
In #183, you mention a SymfonyCommandTask attribute, and wanted to ask about that.

Many of my admin script start out as console commands, I would love to simply tag the command and have it available in castor.  In particular, I don't want to have to re-define the arguments and options, so instead of bin/console app:validate (or symfony console app:validate), I could run castor app:validate [with options, args], without having to explicitly re-define that command.

### PR DESCRIPTION
**Pull request by lyrixx**
fixes #209

cc @tacman and @tucksaun

[simpler diff: `?w=1`](https://github.com/jolicode/castor/pull/254/files?w=1)

### CODE REVIEW

File: doc/06-reference.md [hunk 1]
```diff
@@ -55,3 +55,4 @@ Castor provides the following attributes to register tasks, listener, etc:
 - [`AsListener`](going-further/events.md#registering-a-listener)
 - [`AsOption`](04-arguments.md#overriding-the-option-name-and-description)
 - [`AsTask`](02-basic-usage.md)
+- [`AsSymfonyTask`](going-further/symfony-task.md)
```
Comment (pyrech) [inline doc/06-reference.md:58]: To keep alphabetical order

```suggestion
- [`AsSymfonyTask`](going-further/symfony-task.md)
- [`AsTask`](02-basic-usage.md)
```

File: doc/going-further/symfony-task.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Symfony Task
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:1]: Could you link this new documentation in the index.md? :pray:

File: doc/going-further/symfony-task.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+Symfony application, but you can override it with the `name` parameter.
+
+If Symfony command does own a `AsCommand` attribute, you must set the
+`originName` parameter, and it must be the same as the same in the symfony
+application.
+
+And finally, you can give a way to access the symfony application entry point
+with the `console` parameter. Some examples:
+
+* `['bin/console']`, this is the default, when symfony and castor live in the
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:30]:
```suggestion
* `['bin/console']`, this is the default, when Symfony and Castor live in the
```

File: doc/going-further/symfony-task.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Symfony Task
+
+If castor lives near a Symfony application, you may want to use directly some
+commands. This is possible with the `AsSymfonyTask` attribute:
+
+```php
+#[AsCommand('hello', 'Says hello from a symfony application')]
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:7]:
```suggestion
#[AsCommand('hello', 'Says hello from a Symfony application')]
```

File: doc/going-further/symfony-task.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+If Symfony command does own a `AsCommand` attribute, you must set the
+`originName` parameter, and it must be the same as the same in the symfony
+application.
+
+And finally, you can give a way to access the symfony application entry point
+with the `console` parameter. Some examples:
+
+* `['bin/console']`, this is the default, when symfony and castor live in the
+  very same directory
+* `['path/to/symfony/bin/console']`, when symfony is in another directory
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:32]:
```suggestion
* `['path/to/symfony/bin/console']`, when Symfony is in another directory
```

File: doc/going-further/symfony-task.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+* `['docker', 'exec', 'foobar-backend-1', '/app/server/backend/bin/console']`, when
+  you symfony application lives a docker container
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:34]:
```suggestion
  you Symfony application lives a docker container
```

File: src/Attribute/AsSymfonyTask.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Castor\Attribute;
+
+#[\Attribute(\Attribute::TARGET_CLASS)]
+class AsSymfonyTask
+{
+    /**
+     * @param string[] $console How to start the symfony application
```
Comment (pyrech) [inline src/Attribute/AsSymfonyTask.php:9]:
```suggestion
     * @param string[] $console How to start the Symfony application
```

File: doc/going-further/symfony-task.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
+# Symfony Task
+
+If castor lives near a Symfony application, you may want to use directly some
+commands. This is possible with the `AsSymfonyTask` attribute:
+
+```php
+#[AsCommand('hello', 'Says hello from a symfony application')]
+#[AsSymfonyTask(name: 'symfony:hello', console: [__FILE__])]
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:8]: I think this example could more helpful. IMO it should provide a realistic example (like with `bin/console`))

File: doc/going-further/symfony-task.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+```
+
+By default, the attribute will re-use the same name as the one defined in the
+Symfony application, but you can override it with the `name` parameter.
+
+If Symfony command does own a `AsCommand` attribute, you must set the
+`originName` parameter, and it must be the same as the same in the symfony
+application.
+
+And finally, you can give a way to access the symfony application entry point
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:27]:
```suggestion
And finally, you can give a way to access the Symfony application entry point
```

File: doc/going-further/symfony-task.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# Symfony Task
+
+If castor lives near a Symfony application, you may want to use directly some
+commands. This is possible with the `AsSymfonyTask` attribute:
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:4]:
```suggestion
If your Castor project lives near a Symfony application, you may want to use
some of its commands directly as Castor tasks.
This is possible with the `AsSymfonyTask` attribute you can set on your
command class.
```

> [!NOTE]
> Thanks to how PHP attributes works, your application will work even if it
> does not find this attribute class (which will probably not be available
>  on your vendor directory - unless you installed Castor with Composer).

**Review by pyrech — COMMENTED** [review]
I just left some comments on the documentation but the code looks good to me :yellow_heart:

**Comment by tucksaun** [comment]
I dont' know or use Castor to have a strong opinion about this but code looks legit 🤓 

On the UX side I'm wondering if that makes sense to declare the console entrypoint in your tasks? maybe being able to configure it in `castor.php` would be more practical?

File: doc/going-further/symfony-task.md (hunk 8 above)
Comment (tucksaun) [inline doc/going-further/symfony-task.md:8]: The naming `AsSymfonyTask` makes totally sense in Castor code but that might be weird when you read you Symfony app code... a simple alias could fix this but maybe the doc could use one to simplify the discovery by newcomers? `BindAsCastorTask` for example?

**Comment by tacman** [comment]
Hmm, I've added the AsSymfonyTask to one of my commands, but it's not being recongized.  I'm using castor  v0.11.1, installed manually and using the symfony-command branch.

Is there something else I need to do to the castor.php file?  I re-initialized it, same issue.

```bash
bin/console make:command castor:test
```

Add the attribute
```php
#[AsSymfonyTask(name: 'testing', console: [__FILE__])]
#[AsCommand(
    name: 'castor:test',
    description: 'Add a short description for your command',
)]
```

```bash
castor list
```

```
castor v0.11.1

Usage:
  command [options] [arguments]

Options:
  -h, --help            Display help for the given command. When no command is given display help for the list command
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --ansi|--no-ansi  Force (or disable --no-ansi) ANSI output
  -n, --no-interaction  Do not ask any interactive question
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Available commands:
  completion  Dump the shell completion script
  hello       Welcome to Castor!
  help        Display help for a command
  list        List commands
```

File: doc/going-further/symfony-task.md (hunk 8 above)
Comment (lyrixx) [reply doc/going-further/symfony-task.md:15]: It means we have to embed a whole Symfony project just for one or two commands?
It does not worth it, isn't?

Comment (lyrixx) [reply doc/going-further/symfony-task.md:15]: I Forgot the namespace in the doc, I added it. It's
```
use Castor\Attribute\AsSymfonyTask;
```

It's clearer with it, isn't?

Do you still think it must be changed?

**Comment by lyrixx** [comment]
> I dont' know or use Castor to have a strong opinion about this but code looks legit 🤓

HAN 😱 (joke)

> On the UX side I'm wondering if that makes sense to declare the console entrypoint in your tasks? maybe being able to configure it in `castor.php` would be more practical?

1. we don't have way to configure castor globally
2. It's not possible to configure it globally, since in one repository managed by castor, you can have N Symfony application. So it mush be at the task level. BUT, you can extends the attributes, and preconfigure it, then use this child class to avoid repetition.

**Comment by lyrixx** [comment]
> Is there something else I need to do to the castor.php file? I re-initialized it, same issue.

You need to `import(/path/to/symfony/command/);`

File: doc/going-further/symfony-task.md (hunk 8 above)
Comment (pyrech) [reply doc/going-further/symfony-task.md:22]: Not sure to understand, I believe documentation code sample does not need to be the same as the `examples` directory but must reflect real world usages

Comment (lyrixx) [reply doc/going-further/symfony-task.md:22]: ahhh, got it, this is the documentation! I'll fix it

File: doc/going-further/symfony-task.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+application.
+
+And finally, you can give a way to access the Symfony application entry point
+with the `console` parameter. Some examples:
+
+* `['bin/console']`, this is the default, when Symfony and Castor live in the
+  very same directory
+* `['path/to/symfony/bin/console']`, when Symfony is in another directory
+* `['docker', 'exec', 'foobar-backend-1', '/app/server/backend/bin/console']`, when
+  you Symfony application lives a docker container
```
Comment (pyrech) [inline doc/going-further/symfony-task.md:48]:
```suggestion
  your Symfony application lives in a docker container
```

**Comment by lyrixx** [comment]
>import? I've used require and include, but I'm not sure what import means. the castor import didn't work either. No matter what I do I get

Like all other task, you need to import them. Example with your reproducer

```
import(__DIR__ . '/src/Command');
```

> Cache key "symfony-console-definitions-/home/tac/ca/mus/src/Command/CastorTestCommand.php" contains reserved characters "{}()/@:".

Oups, I'll fix that asap.


>I was hoping that it would do a reflection of all the symfony commands, find the ones that were tagged, and auto-register them.

Yes, that what my PR does :) 

> And of course cache all this so it only did the scan if a file changed, like the CompilerPass of a bundle.

Yes, there is also cache :) 

> A reproducer would be trivial:

I'm on it. Thanks

**Comment by lyrixx** [comment]
I'm not able to reproducer any of the error you reported.

my history, and the diff

```
symfony new --webapp symfony-command-and-castor
cd symfony-command-and-castor
bin/console make:command Hello
# Fix the MAILER_DSN=null 
castor 
code castor.php 
castor hello
```

```diff
diff --git a/.env b/.env
index d7c75ee..5bc2089 100644
--- a/.env
+++ b/.env
@@ -37,5 +37,5 @@ MESSENGER_TRANSPORT_DSN=doctrine://default?auto_setup=0
 ###< symfony/messenger ###
 
 ###> symfony/mailer ###
-# MAILER_DSN=null://null
+MAILER_DSN=null://null
 ###< symfony/mailer ###
diff --git a/castor.php b/castor.php
new file mode 100644
index 0000000..851ed20
--- /dev/null
+++ b/castor.php
@@ -0,0 +1,9 @@
+<?php
+
+use Castor\Attribute\AsTask;
+
+use function Castor\io;
+use function Castor\capture;
+use function Castor\import;
+
+import(__DIR__ . '/src/Command');
diff --git a/src/Command/HelloCommand.php b/src/Command/HelloCommand.php
new file mode 100644
index 0000000..445605a
--- /dev/null
+++ b/src/Command/HelloCommand.php
@@ -0,0 +1,51 @@
+<?php
+
+namespace App\Command;
+
+use Castor\Attribute\AsSymfonyTask;
+use Symfony\Component\Console\Attribute\AsCommand;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+#[AsCommand(
+    name: 'Hello',
+    description: 'Add a short description for your command',
+)]
+#[AsSymfonyTask()]
+class HelloCommand extends Command
+{
+    public function __construct()
+    {
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->addArgument('arg1', InputArgument::OPTIONAL, 'Argument description')
+            ->addOption('option1', null, InputOption::VALUE_NONE, 'Option description')
+        ;
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $io = new SymfonyStyle($input, $output);
+        $arg1 = $input->getArgument('arg1');
+
+        if ($arg1) {
+            $io->note(sprintf('You passed an argument: %s', $arg1));
+        }
+
+        if ($input->getOption('option1')) {
+            // ...
+        }
+
+        $io->success('You have a new command! Now make it your own! Pass --help to see your options.');
+
+        return Command::SUCCESS;
+    }
+}
```

**Comment by tacman** [comment]
here's my reproducer:

```bash
symfony new castor-command-demo --webapp && cd castor-command-demo
sed -i "s|# MAILER_DSN|MAILER_DSN|" .env
bin/console make:command app:castor-test
cat > castor.php <<'END'
<?php

use Castor\Attribute\AsTask;

use function Castor\io;
use function Castor\capture;
use function Castor\import;

import(__DIR__ . '/src/Command/CastorTestCommand.php');

#[AsTask(description: 'Welcome to Castor!')]
function hello(): void
{
    $currentUser = capture('whoami');

    io()->title(sprintf('Hello %s!', $currentUser));
}
END


```

Add attribute to AppCastorTest.php

```php
#[\Castor\Attribute\AsSymfonyTask()]
```

```bash
castor list
```

**Comment by lyrixx** [comment]
Thanks for the reproducer.

1. I didn't get an error about the cache. Are you sure your vendor are up-to-date ? `composer i` ; anyway, I will-ensure we don't put such char in the key
2. I found a cache issue ! I'll fix it ASAP
3. Otherwise, everything is working as expected

**Comment by lyrixx** [comment]
@tacman I pushed a new version. Can you tried it?

Note, you may have to clear the cache manually `rm ~/.cache/castor/ -rf` on my computer

**Review by pyrech — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**